test/cli/fixtures/test-if.js

```diff
@@ -0,0 +1,37 @@
+QUnit.test.if('skip me', false, function (assert) {
+  assert.true(false);
+});
+
+QUnit.test.if('keep me', true, function (assert) {
+  assert.true(true);
+});
+
+QUnit.test('regular', function (assert) {
+  assert.true(true);
+});
+
+QUnit.test.if.each('skip dataset', false, ['a', 'b'], function (assert, _data) {
```

Comment:
These many parameters, including boolean ones, don't read that well. But maybe it's not a big issue in practice since you'd normally see a condition here, not a direct boolean value.

Reply:
Yeah, I'm not super happy with the complexity that this allows.

To me it is a trade-off for in favour of consistency and liberty. It makes sense to have `test.if`. And given the shape of the API, it would imho be expected and least surprising if the API looks the same as for any other `test.*` aliases like `test.skip`, `test.todo`, and `test.only`. Each of those has `.each()` so it makes sense to expose it here as well for consistency.

The downside is that if this freedom is misused, it can create some pretty complex tests. I've mitigated that by not explicitly demonstrating or encouraging this edge case in documentation. I think in most cases the way you reach this is naturally over time. When taken in small steps, I imagine developers would expect this to exist, which seems preferred over the surprise and cognitive complexity that would arise from us judging and omitting it in this one case. It remains of course, as always, the responsibility of developers and code review to judge their apetite for complexity, and what the alternative is.

The alternative in this case might be commenting out (reducing test coverage) or wrapping in an if-statement without leaving a trace of the skipped test.

Another alternative might be to use this extreme example as evidence that the API is too complex, and that this complexity even hides but exists in the simpler case of `QUnit.test.if(name, cond, fn)`. If we believe that, we could move toward an API like `assert.skip()` where you'd have to write it like a normal test:

```js
QUnit.test('example', function (assert) {
  if (cond) {
    assert.skip(); // throws special internal exception, stops test, marks it as skipped.
  }
});
```

I tend to dislike this because it's encouraging and normalizing use of `assert` in nested blocks. It also departs from the QUnit model where assertion methods are generally not test-ending. It also creates ambiguity over what to do with assertions that may happen before this (do we throw an Error to say that's invalid?). And whether to have beforeEach/afterEach. Even if there are no assertions, there may be other setup code in it, that would warrant before/afterEach. And in fact you may want to use some of your before/after code to power the conditional. That's all stuff we avoid with the declarative `test.if` model, at the cost of a longer method name, and longer signature.